File: aoj/problems/0529/0529.cpp
///
// File:  0529.cpp
// Author: ymiyamoto
//
// Created on Wed Feb 28 21:12:52 2018
//
#include <bits/stdc++.h>
using namespace std;

#define rep(_n, _init, _N) for (int32_t _n = _init; _n < (int32_t)(_N); _n++)
#define rrep(_n, _N, _end) for (int32_t _n = (int32_t)(_N); _n >= _end; _n--)
#define iceil(_x, _y) (((_x) + (_y)-1) / (_y))
#define ifloor(_x, _y) ((_x) / (_y))

static const double pi = 2 * asin(1);
static const double precision = 10e-9;

int32_t main()
{
  while (true) {
    int64_t N, M;
    cin >> N >> M;
    if (N == 0) break;

    vector<int64_t> arr = vector<int64_t>(N + 1);
    rep(i, 1, N + 1)
    {
      cin >> arr[i];
    }
    sort(arr.begin(), arr.end());

    vector<int64_t> sums = vector<int64_t>((N + 1) * (N + 1));
    uint32_t cnt = 0;
    rep(i, 0, N + 1)
    {
      rep(j, 0, N + 1)
      {
        sums[cnt] = arr[i] + arr[j];
        cnt++;
      }
    }

    sort(sums.begin(), sums.end());
    int64_t m = 0;
    rep(i, 0, sums.size())
    {
      int64_t rem = M - sums[i];
      if (0 <= rem) {
        auto p = lower_bound(sums.begin(), sums.end(), rem);
        if (p != sums.begin() && p != sums.end()) {
          m = max(m, *(p - 1) + sums[i]);
        }
      }
    }
    cout << m << endl;
  }

  return 0;
}

File: aoj/problems/0529/0529.go
///
// File:  0529.go
// Author: ymiyamoto
//
// Created on Wed Feb 28 20:39:28 2018
//
package main

import (
	"fmt"
	"sort"
)

func max(n, m int64) int64 {
	if n < m {
		return m
	} else {
		return n
	}
}

func main() {
	for {
		var N int
		var M int64
		fmt.Scan(&N, &M)
		if N == 0 && M == 0 {
			break
		}

		arr := make([]int64, N+1)
		for i := 1; i <= N; i++ {
			fmt.Scan(&arr[i])
		}
		sort.Slice(arr, func(i, j int) bool { return arr[i] < arr[j] })

		sums := make([]int64, len(arr)*len(arr))
		for i := 0; i < len(arr); i++ {
			for j := 0; j < len(arr); j++ {
				sums[i*len(arr)+j] = arr[i] + arr[j]
			}
		}

		sort.Slice(sums, func(i, j int) bool { return sums[i] < sums[j] })
		var m int64 = 0
		for _, val := range sums {
			rem := M - val
			if rem >= 0 {
				index := sort.Search(len(sums), func(i int) bool { return sums[i] > rem }) - 1
				if index >= 0 {
					m = max(m, val+sums[index])
				}
			}
		}
		fmt.Println(m)
	}
}
